kernel/src/capability_ptr.rs

```diff
@@ -0,0 +1,138 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Google LLC 2024.
+
+//! Defines the CapabilityPtr type
+
+use core::fmt::{Formatter, LowerHex, UpperHex};
+use core::ops::AddAssign;
+
+/// A pointer with target specific metadata concerning validity or access rights.
```

Comment:
It's not clear to me how we could actually use this type in upstream Tock since we don't have hardware that supports this metadata.

I still think we need an abstraction which captures the intent and not the implementation. This is tied to the naming question, but is distinct. It may actually make sense to call _this_ struct `CapabilityPtr` given this comment description. However, I think something like `QualifiedPtr` better captures the distinction between the type of pointer considered in this PR, and other types of pointers. A `QualifiedPtr` would be described entirely based on its use case within Tock, and not at all based on how it could/is implemented.

Reply:
I don't expect this type to add much value over `*const ()` in upstream Tock, as I think it only makes sense on the platforms  that have such a thing, so I don't think it will have any added benefit per se. 

However, I think Capability does carry some semantic abstraction without just being a CHERI capability. For example, if Tock ever targeted a platform with PAC (which I would frame as password capabilities), I could imagine the checked parts of this interface checking pointer integrity, casts from usize signing the pointer etc.

I think that "authority bearing" does convey what it is used for and where it should be used and is independent of implementation details, such as how that authority is encoded or checked.

Reply:
I think there is some talking past each other here.

I don't think @bradjc is looking for "value" in non-CHERI Tock, but rather an explanation/name/whatever of the abstraction (the thing called `CapabilityPtr`) that makes it clear how it should be used. E.g., it is describing some concept. That concept is _enforced_ if there are hardware capabilities. It's not enforced if there are not.  But it's still meaningful to _use_ that abstraction elsewhere, if only as a description of the meaning of some particular value.

My working model is something like: a `CapabilityPtr` names some address in userspace (though not necessarily currently/always accessible to userspace) that, conceptually, userspace may or may not be permitted to read, write, or execute. There are certain operations on this abstraction, and those operations carry some implicit (enforced or otherwise) affect on those permissions. For example, I can add/subtract offsets from a `CapabilityPtr`, and the result may or may not have the same access permissions (e.g. the permissions will drop if the offset puts me out of bounds of the region it referred to). I can also extract the word-size address component as pure data (a `usize`), which let's me "hide" data inside a `CapabilityPtr` in the same way I can "hide" data in a `*const _`.

Again, these properties may or may not be enforced. In this sense it's kind of like the difference between a bounds-checked slice in Rust and an array in C with no bounds checking. In all cases, correct programs will behave the same, and will use these in accordance with the rules/properties. On CHERI, an incorrect program will fault if it uses it incorrectly. Similarly, a Tock kernel that uses this concept correctly in the right places will enable correct userspace programs regardless, but a Tock kernel that uses it _incorrectly_ will only fail to support programs under CHERI (because in non-CHERI userspace programs can also de-reference a `usize`/`*const _` so the type that's passed up from the kernel doesn't actually matter).

OK, this _concept_ now tells us where/how to use a `CapabilityPtr`. For example, if a system call is intending to return a value that might mean "a pointer userspace may dereference in some way" it should be a `CapabilityPtr` (e.g. the user-data passed down in the allow system call and up in upcalls). If a value is passed between userspace and the kernel that should _never_ be treated as a pointer userspace can dereference, it should not be a `CapabilityPtr`.

(perhaps I'm wrong about the specific explanation of the abstraction, but hopefully this at least captures the gap in understanding)

Reply:
I think about this abstraction in a similar way. Maybe then, I am not sure   what the suggestion for change is here. Just that `Capability` is still not a good name for this thing? Or that the text could do a better job explaining it use? Or its meaning?

Reply:
> Or that the text could do a better job explaining it use? Or its meaning?

@bradjc, confirm?

I _think_ the clarifying the meaning is necessary to review whether the rest of the PR is indeed using it in an appropriate way, etc.  Especially to the extent it concerns things like changing TRD104 and related core kernel interfaces.

If that's the case, I can also take a stab at writing this out. I can see how the comment explaining this below is maybe too CHERI and implementation oriented

Reply:
Amit's comment is spot on, and the changes in https://github.com/tock/tock/pull/4174/commits/7ec6bdd51152cc054c2f68dcab3425928ad11876 I think make it a lot more clear what role this serves in Tock. I'd even like to see the entire comment included in that file.

Reply:
> However, I think Capability does carry some semantic abstraction without just being a CHERI capability. For example, if Tock ever targeted a platform with PAC (which I would frame as password capabilities), I could imagine the checked parts of this interface checking pointer integrity, casts from usize signing the pointer etc.

I think one of the questions I have is, in this scenario, would PAC use the _same_ `CapabilityPtr` as-defined here and the same syscall interface to pass the capability pointer, etc?

More important is up-leveling from that specific question and away from something specific to CHERI or to PAC or to [thing]—is what's proposed here a _general mechanism_ for conveying pointers with metadata? If so, how do we express what the properties of a "pointer with metadata" abstraction should be?

A related question based on @alevy's new text, is this _only_ for userspace pointers? Are theses capabilities _only_ enforced during userspace operation, or are there cases where we would want to use hardware capability enforcement to e.g. support stronger isolation / IFT / etc between resources passed to capsules? Not to re-bikeshed the name, but if it's really userspace-only, that feels like it should be a front a center piece of the naming/rationale.

I'm also struggling with the repeated assertion that this is _only_ useful on platforms with hardware capabilities. I feel like that isn't the case? i.e., a non-C userspace with some kind of managed runtime could in principle [less-efficient, sure] enforce all the same properties as CHERI or equivalent hardware. Perhaps explaining the pointer abstraction here with the perspective of also answering "what would a pure-software implementation that intercepts every memory access provide" would help, as it would divorce the discussion from any particular implementation and invite thinking of things in terms of the more fundamental properties?

Reply:
> Would PAC use the same CapabilityPtr as-defined here and the same syscall interface to pass the capability pointer, etc

Hopefully. It may have to grow to assume a superset of hardware checked integrity (CHERI motivated bounds and W,R,X, but one can imagine more).

> A related question based on @alevy's new text, is this only for userspace pointers? 

No, I have left a comment there. The boundary is just the lowest hanging fruit because it is the boundary where compiler enforcement stops applying and we must rely on hardware.

> i.e., a non-C userspace with some kind of managed runtime could in principle [less-efficient, sure] enforce all the same properties as CHERI or equivalent hardware

Well, I think it does, and I think that language is `Rust` and that is already precisely what libtock-rs does. I would argue that, in the rust type system, `&T` is _already_ a capability enforced by the type system, and libtock-rs enforces that you might syscalls only when you have one of those. The point of `CapabilityPtr` was to be a source for authority beyond the existing type system (which is why it immediately found use across compiler boundaries). We cannot proves you used libtock-rs properly. Use of `CapabilityPtr` can fix issues in even Rust: when either the compiler is wrong, or unsafe Rust invokes UB. Unsafe Rust can fabricate `&T`. In userspace or the kernel. On some platforms (namely CHERI), it cannot fabricate a valid CapabilityPtr.

Reply:
> is what's proposed here a general mechanism for conveying pointers with metadata?

@ppannuto I intentionally avoided defining this in terms of metadata. It has a semantic meaning relating to user references. On some platforms thateaning might be enforcible with hardware support, on other it won't.

The important bit is that this type is a (valid or invalid) userspace reference. That there may be metadata on some platforms is incidental.

Reply:
What does PAC stand for?

Reply:
Pointer Authentication Code: https://learn.arm.com/learning-paths/servers-and-cloud-computing/pac/pac/